### PR TITLE
Integrate container testing for Centos

### PR DESCRIPTION
**Pull request by b10n1k**
Difference with https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/11166 is that i add test data variable instead of job variable.

- Related ticket: https://progress.opensuse.org/issues/66595
- [Verification run](https://openqa.suse.de/tests/overview?version=15-SP3&distri=sle&build=b10n1k%2Fos-autoinst-distri-opensuse%2311201)

### CODE REVIEW

File: tests/containers/docker_image.pm [hunk 1]
```diff
@@ -15,42 +15,38 @@
...
-use version_utils qw(is_sle is_opensuse is_tumbleweed is_leap is_public_cloud);
+use scheduler 'get_test_suite_data';
 
 sub run {
-    my ($self) = @_;
-    $self->select_serial_terminal;
-
     my ($image_names, $stable_names) = get_suse_container_urls();
+    my $test_data           = get_test_suite_data();
+    my $container_mgmt_tool = "docker";
```
Comment (jlausuch) [inline tests/containers/docker_image.pm:31]: is this really needed? This module is called "docker". I would imagine this is useful if we had a  module which is used for both docker and podman, but in the case we have, everything is docker inside this .pm file.

File: tests/containers/import_ca_certs.pm [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Make sure that ca certifications were installed
+# Maintainer: qa-c team <qa-c@suse.de>
+
+use Mojo::Base qw(consoletest);
+use testapi;
+use utils;
+use containers::common;
```
Comment (jlausuch) [inline tests/containers/import_ca_certs.pm:16]: maybe we don't need this lib.

File: tests/containers/import_ca_certs.pm [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Make sure that ca certifications were installed
+# Maintainer: qa-c team <qa-c@suse.de>
+
+use Mojo::Base qw(consoletest);
+use testapi;
+use utils;
```
Comment (jlausuch) [inline tests/containers/import_ca_certs.pm:15]: do we need `utils`?

File: tests/containers/setup_centos.pm [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+# Maintainer: qa-c team <qa-c@suse.de>
+
+use Mojo::Base qw(consoletest);
+use testapi;
+
+sub run {
+    my ($self) = @_;
+    $self->select_serial_terminal;
+    assert_script_run "dhclient -v";
+    assert_script_run "curl http://ca.suse.de/certificates/ca/SUSE_Trust_Root.crt -o /etc/ssl/certs/SUSE_Trust_Root.crt";
```
Comment (jlausuch) [inline tests/containers/setup_centos.pm:22]: Could we run the module `containers/import_ca_certs` in centos as well instead of this line? OR at least moving the logic of this certificates to that module instead of this one?

**Comment by jlausuch** [comment]
I prefer this way, everything is in the same place, so the logic is not duplicated. Good job.
Just a few comments.

File: lib/containers/common.pm [hunk 5]
```diff
@@ -28,35 +28,49 @@ our @EXPORT = qw(install_podman_when_needed install_docker_when_needed allow_sel
...
 
 sub install_docker_when_needed {
-    if (is_microos) {
-        # Docker should be pre-installed in MicroOS
-        die 'Docker is not pre-installed.' if zypper_call('se -x --provides -i docker');
+    my $host_os = shift;
+
+    if ($host_os == 'centos') {
+        assert_script_run "dnf config-manager --add-repo=https://download.docker.com/linux/centos/docker-ce.repo";
+        # if podman installed allowerasing to solve conflicts
```
Comment (asmorodskyi) [inline lib/containers/common.pm:54]:
```suggestion
        # if podman installed allow erasing to solve conflicts
```

Comment (asmorodskyi) [reply lib/containers/common.pm:54]: or if you wanted to put exact flag name than something like 
```suggestion
        # if podman installed use flag "--allowerasing" to solve conflicts
```

current version looks like typo , it is not obvious that you meant flag name

File: tests/containers/import_ca_certs.pm (hunk 2 above)
Comment (b10n1k) [reply tests/containers/import_ca_certs.pm]: you are right. removed

File: tests/containers/import_ca_certs.pm (hunk 3 above)
Comment (b10n1k) [reply tests/containers/import_ca_certs.pm:15]: This actually is used and is needed

File: lib/containers/common.pm (hunk 5 above)
Comment (b10n1k) [reply lib/containers/common.pm]: Done

File: lib/containers/common.pm [hunk 6]
```diff
@@ -28,32 +28,48 @@ our @EXPORT = qw(install_podman_when_needed install_docker_when_needed allow_sel
...
+    my @pkgs = qw(podman);
     if (script_run("which podman") != 0) {
-        if (is_sle '>=15') {
-            add_suseconnect_product('sle-module-containers');
+        if ($host_os eq 'centos') {
+            assert_script_run "dnf -y install @pkgs", timeout => 160;
+        }
+        else {
+            if (is_sle '>=15') {
+                add_suseconnect_product('sle-module-containers');
```
Comment (jlausuch) [inline lib/containers/common.pm:40]:
```suggestion
                add_suseconnect_product('sle-module-containers') if (is_sle '>=15');
```

File: lib/containers/common.pm [hunk 7]
```diff
@@ -28,32 +28,48 @@ our @EXPORT = qw(install_podman_when_needed install_docker_when_needed allow_sel
   test_container_runtime test_container_image scc_apply_docker_image_credentials scc_restore_docker_image_credentials);
 
 sub install_podman_when_needed {
+    my $host_os = shift;
+    record_info $host_os;
```
Comment (jlausuch) [inline lib/containers/common.pm:32]:
```suggestion
    my $host_os = shift //= 'sle';
    record_info('HOST', $host_os);
```

File: lib/containers/container_images.pm [hunk 8]
```diff
@@ -95,7 +95,7 @@ sub test_opensuse_based_image {
 
     my $distri  = $args{distri}  //= get_required_var("DISTRI");
     my $version = $args{version} //= get_required_var("VERSION");
-
+    my $host_os = $args{host_os};
```
Comment (jlausuch) [inline lib/containers/container_images.pm:98]: why do you need this?  I don't see you use this variable in this function.

File: lib/containers/container_images.pm [hunk 9]
```diff
@@ -129,4 +129,11 @@ sub test_opensuse_based_image {
     assert_script_run("$runtime run --rm refreshed-image sh -c 'zypper -v ref | grep \"All repositories have been refreshed\"'", 120);
 }
 
+sub exec_on_container {
+    my ($image, $runtime, $command, $timeout) = @_;
+    $timeout //= 120;
+    assert_script_run("$runtime run --rm $image $command", $timeout);
+    save_screenshot;
```
Comment (jlausuch) [inline lib/containers/container_images.pm:136]: need to save screenshot?

File: tests/containers/docker_image.pm (hunk 1 above)
Comment (jlausuch) [reply tests/containers/docker_image.pm:31]: yeah, I like the idea of making things generic when it makes sense, but then we still have specific functions in this module like `install_docker_when_needed` and `scc_apply_docker_image_credentials `

Comment (jlausuch) [reply tests/containers/docker_image.pm:31]: Maybe not for this PR, but an idea that comes to my mind out of this discussion is to have `docker_image` and `podman_image` as a single file with your idea of `$container_mgmt_tool` variable, because the logic of both modules are basically the same.

File: tests/containers/setup_centos.pm [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+# - import SUSE CA certificates
+# Maintainer: qa-c team <qa-c@suse.de>
+
+use Mojo::Base qw(consoletest);
+use testapi;
+
+sub run {
+    my ($self) = @_;
+    $self->select_serial_terminal;
+    assert_script_run "dhclient -v";
```
Comment (jlausuch) [inline tests/containers/setup_centos.pm:21]: I agree in having modules to setup specific Host OS, but in this case it's only to run 1 command... I wonder if it would be more efficient moving this command to `install_docker_when_needed.pm` in the condition for centos.

File: lib/containers/common.pm (hunk 6 above)
Comment (b10n1k) [reply lib/containers/common.pm:40]: That was not my code but i will adjust it. more perlist

File: lib/containers/common.pm (hunk 7 above)
Comment (b10n1k) [reply lib/containers/common.pm:32]: I forgot to remove the recored_info. But the default is not need it here. the default is sle/opensue which it will run when it reach the `else`

File: lib/containers/container_images.pm (hunk 9 above)
Comment (b10n1k) [reply lib/containers/container_images.pm:136]: forgot to remove

File: tests/containers/setup_centos.pm (hunk 10 above)
Comment (b10n1k) [reply tests/containers/setup_centos.pm:21]: i dont like it too. But also it does not belong to other places. So one idea was to rename the module to reflect to the general setup. As this will possible include configuration for ubuntu and alpine later. But also i could merge it with the `import_ca_certs.pm`

File: lib/containers/common.pm (hunk 6 above)
Comment (jlausuch) [reply lib/containers/common.pm:40]: I know :)  But since you are improving this tests, maybe some small thing to do :)

File: tests/containers/setup_centos.pm (hunk 10 above)
Comment (jlausuch) [reply tests/containers/setup_centos.pm:21]: Up to you, I don't mind if we have a module for each 3rd party host setup. Let's see what others think.

Comment (b10n1k) [reply tests/containers/setup_centos.pm]: i made some changes here af17ee2de

File: lib/containers/container_images.pm (hunk 8 above)
Comment (b10n1k) [reply lib/containers/container_images.pm]: forgotten. has been removed. i was thinking to add some logic here which i did not need at the end

File: tests/containers/docker_image.pm (hunk 1 above)
Comment (b10n1k) [reply tests/containers/docker_image.pm:31]: i noticed too. But i did not want to do much refactoring on this PR. There was also a [comment](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/11166/files#r505296504) in the previous PR to approach things in OO. I would like the idea but i do not know how far you want to go. Let s discuss it outside the ticket

File: tests/containers/docker_image.pm [hunk 11]
```diff
@@ -15,42 +15,38 @@
...
+            test_opensuse_based_image(image => $iname, runtime => $container_mgmt_tool);
+            build_with_zypper_docker(image => $iname, runtime => $container_mgmt_tool);
+            scc_restore_docker_image_credentials() if (get_var('SCC_DOCKER_IMAGE'));
+        }
+        else {
+            exec_on_container($iname, $container_mgmt_tool, 'cat /etc/os-release');
+        }
     }
-
-    scc_restore_docker_image_credentials() if (get_var('SCC_DOCKER_IMAGE'));
```
Comment (pdostal) [inline tests/containers/docker_image.pm:52]: Why is this line removed :(

File: tests/containers/host_configuration.pm [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+use utils;
+use scheduler 'get_test_suite_data';
+
+sub run {
+    my ($self) = @_;
+    $self->select_serial_terminal;
+    my $test_data = get_test_suite_data();
+    assert_script_run "dhclient -v" if $test_data->{host_os} eq 'centos';
+    ensure_ca_certificates_suse_installed() unless $test_data->{host_os};
+    assert_script_run "curl http://ca.suse.de/certificates/ca/SUSE_Trust_Root.crt -o /etc/ssl/certs/SUSE_Trust_Root.crt";
```
Comment (pdostal) [inline tests/containers/host_configuration.pm:27]: I don't understand this. I'd do something like:
```perl
if (is_sle() || is_opensuse()) {
    ensure_ca_certificates_suse_installed() unless $test_data->{host_os};
} else {
    assert_script_run "curl http://ca.suse.de/certificates/ca/SUSE_Trust_Root.crt -o /etc/ssl/certs/SUSE_Trust_Root.crt";
}
```

Comment (b10n1k) [reply tests/containers/host_configuration.pm:27]: Think of it as "if `$test_data->{host_os}` is other than is_sle/is_openesuse" The extra if/else do not make any difference here. `is_sle` and `is_opensuse` will be always true (unless jeos)

File: tests/containers/docker_image.pm (hunk 11 above)
Comment (b10n1k) [reply tests/containers/docker_image.pm:52]: unintentional. restored

File: tests/containers/host_configuration.pm (hunk 12 above)
Comment (jlausuch) [reply tests/containers/host_configuration.pm:27]: what about the `curl` command? I guess we want to execute that only in non-suse OSs, right?

File: tests/containers/docker_image.pm (hunk 1 above)
Comment (jlausuch) [reply tests/containers/docker_image.pm:31]: Sure, we can do this in another PR.

File: tests/containers/host_configuration.pm (hunk 12 above)
Comment (b10n1k) [reply tests/containers/host_configuration.pm]: That sounds right. I will adjust.

File: tests/containers/docker_image.pm (hunk 1 above)
Comment (pdostal) [reply tests/containers/docker_image.pm:31]: I would keep `'docker'` until we decide to merge the `docker_image.pm` and `podman_image.pm` together.

Comment (b10n1k) [reply tests/containers/docker_image.pm:31]: Maybe i do not understand this correctly. The string docker appears in many statements in the file. So i assigned it to a variable. is the problem the name of the variable?

Comment (pdostal) [reply tests/containers/docker_image.pm:31]: Probably yes. Maybe use `$runtime` instead of `$container_mgmt_tool` so it is consistent.

**Comment by pdostal** [comment]
You inroduce `$test_data->{host_os}`.
Why don't you set `DISTRI=centos` instead?

**Comment by asmorodskyi** [comment]
> You inroduce `$test_data->{host_os}`.
> Why don't you set `DISTRI=centos` instead?

you can not do this . `DISTRI` variable used by openQA backend to figure out where main.pm and needles are

**Comment by pdostal** [comment]
> > You inroduce `$test_data->{host_os}`.
> > Why don't you set `DISTRI=centos` instead?
> 
> you can not do this . `DISTRI` variable used by openQA backend to figure out where main.pm and needles are

It is one simlink. We already do this for `sle`.

**Comment by b10n1k** [comment]
http://aquarius.suse.cz/tests/3571#
We will have to have a symlink to centos in this case. As it is now the /var/lib/openqa/share/tests/centos does not exist. Do we want this for a single scenario. And consider that we have tasks to add ubuntu and alpine

**Comment by jlausuch** [comment]
> > > You inroduce `$test_data->{host_os}`.
> > > Why don't you set `DISTRI=centos` instead?
> > 
> > 
> > you can not do this . `DISTRI` variable used by openQA backend to figure out where main.pm and needles are
> 
> It is one simlink. We already do this for `sle`.

not only a symlink, it should have its main.pm and needles directory. In JeOS, we use DISTRI=sle because of not over complicating this too much and reusing the same main and sle directory. For just 1 test I would keep DISTRI=sle and use another way to detect Host OS.

**Comment by pdostal** [comment]
OK, just for the record the plan is to remove `$test_data` stuff and check `/etc/os-release` instead.

File: tests/containers/docker_image.pm (hunk 1 above)
Comment (b10n1k) [reply tests/containers/docker_image.pm]: Done

**Review by pdostal — APPROVED** [review]
I'm goo! Thank you @b10n1k

**Review by jlausuch — APPROVED** [review]
